docs/modules/otaupgrade.md

```diff
@@ -0,0 +1,176 @@
+# BT HCI Module
+| Since  | Origin / Contributor  | Maintainer  | Source  |
+| :----- | :-------------------- | :---------- | :------ |
+| 2019-06-24 | [DiUS](https://github.com/DiUS), [Johny Mattsson](https://github.com/jmattsson) | [Johny Mattsson](https://github.com/jmattsson) | [otaupgrade.c](../../components/modules/otaupgrade.c)|
+
+The OTA Upgrade module provides access to the IDF Over-The-Air Upgrade
+support, enabling new application firmware to be applied and booted into.
+
+This module is not concerned with where the new application comes from.
+The choice of download source and method (e.g. https, tftp) is left to
+the user.
+
+In order to use the `otaupgrade` module, there must exist at least two
+OTA partitions (type `app`, subtype `ota_0` / `ota_1`), as well as the
+"otadata" partition (type `data`, subtype `ota`). The IDF implements
+the typical "flip-flop" approach to upgrades, in that one of the
+partitions hosts the running application, and the upgrade is downloaded
+into the inactive partition and only when fully downloaded and verified
+is it marked as bootable. This makes the system resilient to incomplete
+upgrades, be it due to power-loss, interrupted downloads, or other such
+things.
+
+Depending on whether the installed boot loader has been built with or
+without rollback support, the upgrade process has four or three steps.
+Without rollback support, the steps are:
+
+- `otaupgrade.commence()`
+- feed the new application image into `otaupgrade.write(data)` in chunks
+- `otaupgrade.complete(1)` to finalise and reboot into the new application
+
+If the boot loader is built with rollback support, an extra step is needed
+after the new application has booted (and been tested to be "good", by
+whatever metric(s) the user chooses):
+
+- `otaupgrade.accept()` to mark this image as valid, and allow it to be
+  booted into again.
+
```

Comment:
Is that to say that if the module reboots from the new firmware without accepting (and without paving over the old firmware) it will load the old version?  Does the IDF layer prevent a not-yet-accepted firmware from OTAing over the old firmware?

Reply:
Yes, and yes. If you try to blatter over the old image before you've marked the new, running one as accepted, the `commence()` will fail.